### DISCUSSION TITLE
Question on Process List Query Priority

### DISCUSSION THREAD

**Discussion by msk** [aicers/review-web#158] [Q&A] [answered]
I've been working on #144 and came across an interesting piece of implementation in the `processList` query function. The code in question prioritizes certain agents over others when requesting a list of processes for a given host.

```rust
let priority_app = apps
    .iter()
    .find(|(x, _)| x == &format!("hog@{hostname}"))
    .or_else(|| {
        apps.iter()
            .find(|(x, _)| x == &format!("piglet@{hostname}"))
    })
    .or_else(|| {
        apps.iter()
            .find(|(x, _)| x == &format!("crusher@{hostname}"))
    });
```

I noticed that the agents `hog`, `piglet`, and `crusher` are prioritized in that specific order. @BLYKIM, could you shed some light on the decision behind this prioritization? Are there specific performance or reliability advantages to this approach? I'm curious if targeting agents in this manner influences the efficiency or effectiveness of the `processList` query.

**Comment by BLYKIM** [comment]
안녕하세요, @msk 
우선, process_list를 반환하는 기능은 review_web 자체 서버가 아니라면 hog, piglet, crusher 세군데에 구현되었습니다. (https://github.com/aicers/roxy/issues/235)
구현 당시 hog를 dummy_hog로 사용할 경우, review와의 통신을 통해 서버의 역할을 구분하지 않고 서버의 상태만을 반환해주는 역할로 사용하였기 때문에 최우선순위로 두었습니다.
dummy_hog를 사용하지 않는 서버의 경우, 각 서버를 담당하는 piglet이나 crusher를 통해 process_list를 전달 받으며, hog의 후순위에는 특별한 이유는 없었던 것 같습니다. piglet이냐 crusher이냐에 대한 순위는 동시에 찾도록 수정되어도 될 것 같습니다.

다만 tis프로젝트로 구현하고 오랜만에 보는 것 이기에 놓치는 점이 있을 수도 있습니다. @syncpark 추가 확인 부탁드립니다. 감사합니다.

**Comment by syncpark** [comment, marked as answer]
In some project, there are servers that REview must monitor it's status but our softwares are not installed.
We thought `Hog` without input event (i.e., dummy hog) can take that role. 
That's why the `Hog` have the highest priority.

`REconverge` can not be used to collect process list, because it does not maintain it's connection to REview continuously.
`Giganto` can not do that either, because Giganto does not have QUIC connection to REview and `Crusher or  dummy Hog` can do that role instead of `Giganto`.

We came to the conclusion that we could monitor all servers with just Crusher, Piglet, and Hog.
Considering the load of above 3 programs, the order of Hog, Crusher, and Piglet is best.

**Reply by msk** [reply]
Thank you for the detailed explanation.

Based on your explanation, I plan to add the following comment to the code as part of my work on #144, which will touch this part of the code. This will ensure that the rationale behind this prioritization is clear to anyone working on or reviewing the code in the future:

```rust
// The priority of applications (`Hog`, followed by `Crusher` and
// then `Piglet`) is determined based on their roles and
// capabilities in the system.
//
// 1. `Hog` has the highest priority because it can act as a dummy
//    agent to monitor servers where our software isn't installed.
//    This is essential for comprehensive system monitoring.
// 2. `Crusher` and `Piglet` are preferred for process list
//    collection because they maintain a continuous connection to
//    REview.
// 3. `REconverge` is not suited for process list collection due to
//    its intermittent connection.
// 4. The order of Hog, Crusher, and Piglet also considers the load
//    each program can handle.
```
